### PR TITLE
World generators

### PR DESCRIPTION
**Pull request by Antifluxfield**
An improved version of old generators.
All generators are generally finished.
Tested for a few times, and no obvious issue found.
Ore veins now generated per chunk; stones and asteroids don't. I tried once but failed, and not yet find out what's wrong...

### CODE REVIEW

**Comment by SuperCoder7979** [comment]
along with asteroids, would it be possible to spawn ores on the new infinite end generation? Vanilla has that feature we might as well use it :D

**Comment by Antifluxfield** [comment]
@SuperCoder7979 Yes, they do.

**Comment by SuperCoder7979** [comment]
Awesome!

File: src/main/java/gregtech/common/worldgen/GTWorldGen_Asteroid.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import gregtech.api.unification.material.type.DustMaterial;
+import gregtech.api.unification.ore.StoneType;
+import gregtech.api.unification.ore.StoneTypes;
+import gregtech.api.util.WeightedList.WeightedWrapperList;
+import net.minecraft.world.World;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.chunk.IChunkProvider;
+import net.minecraft.world.gen.IChunkGenerator;
+
+public class GTWorldGen_Asteroid extends GTWorldGen_Stone {
```
Comment (Exidex) [inline src/main/java/gregtech/common/worldgen/GTWorldGen_Asteroid.java]: Use Java naming convention

File: src/main/java/gregtech/common/worldgen/GTWorldGenStone.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        for (int i = (minX & ~0xF); i <= maxX; i += 16) {
+            for (int j = (minZ & ~0xF); j <= maxZ; j += 16) {
+                chunkCoords.add(new ChunkPos(i, j));
+            }
+        }
+        chunkCoords/*.parallelStream()*/.forEach(chunkCoord -> {
+            int x0 = Math.max(chunkCoord.x, minX);
+            int x1 = Math.min(chunkCoord.x + 16, maxX);
+            int y0 = Math.max(chunkCoord.z, minZ);
+            int y1 = Math.min(chunkCoord.z + 16, maxZ);
```
Comment (Exidex) [inline src/main/java/gregtech/common/worldgen/GTWorldGenStone.java]: This seams to do nothing

File: src/main/java/gregtech/common/worldgen/GTWorldGenStone.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    @FunctionalInterface
+    protected interface ActionAtPos {
+        /**
+         * @param pos       Position of the block
+         * @param r         Square of the relative "distance" metric of the block to the center area of the stone.
+         *                  Ranged between 0.0f to 1.0f
+         * @param random A local {@link Random}}
+         */
+        public void generate(BlockPos pos, float r, Random random);
```
Comment (Exidex) [inline src/main/java/gregtech/common/worldgen/GTWorldGenStone.java]: `public` on interface methods is redundant

File: src/main/java/gregtech/common/worldgen/WorldGenerator.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            GTWorldGenOreVein.getRandomOreVein(rnd, world, biome).ifPresent(oregen -> oregen.generate(rnd, chunkX, chunkZ, centerX, centerZ, world, biome, chunkGenerator, chunkProvider));
+        } catch (Exception e) {
+            GTLog.logger.catching(e);
+        }
+    }
+
+    private Boolean sorted = false;
+
+    public WorldGenerator() {
+        GameRegistry.registerWorldGenerator(this, 1073741823);
```
Comment (Exidex) [inline src/main/java/gregtech/common/worldgen/WorldGenerator.java]: make constructor private and create static init method like all other init methods

**Pull request merged**